tobac/themes/tobac_v1/segmentation.py

```diff
@@ -0,0 +1,449 @@
+    # from skimage.segmentation import random_walker
+    from scipy.ndimage import distance_transform_edt
+    from copy import deepcopy
+    import numpy as np
+
+    # copy feature dataframe for output 
+    features_out=deepcopy(features_in)
+    # Create cube of the same dimensions and coordinates as input data to store mask:        
+    segmentation_out=1*field_in
+    segmentation_out.rename('segmentation_mask')
+    segmentation_out.units=1
+
+    #Create dask array from input data:
+    data=field_in.core_data()
+    
+    #Set level at which to create "Seed" for each feature in the case of 3D watershedding:
+    # If none, use all levels (later reduced to the ones fulfilling the theshold conditions)
+    if level==None:
+        level=slice(None)
+
+    # transform max_distance in metres to distance in pixels:
+    if max_distance is not None:
+        max_distance_pixel=np.ceil(max_distance/dxy)
+
+    # mask data outside region above/below threshold and invert data if tracking maxima:
+    if target == 'maximum':
+        unmasked=data>threshold
+        data_segmentation=-1*data
+    elif target == 'minimum':
+        unmasked=data<threshold
+        data_segmentation=data
+    else:
+        raise ValueError('unknown type of target')
+
+    # set markers at the positions of the features:
+    markers = np.zeros(unmasked.shape).astype(np.int32)
+    if field_in.ndim==2: #2D watershedding        
+        for index, row in features_in.iterrows():
+            markers[int(row['hdim_1']), int(row['hdim_2'])]=row['feature']
+
+    elif field_in.ndim==3: #3D watershedding
+        list_coord_names=[coord.name() for coord in field_in.coords()]
+        #determine vertical axis:
+        if vertical_coord=='auto':
+            list_vertical=['z','model_level_number','altitude','geopotential_height']
+            for coord_name in list_vertical:
+                if coord_name in list_coord_names:
+                    vertical_axis=coord_name
+                    break
+        elif vertical_coord in list_coord_names:
+            vertical_axis=vertical_coord
+        else:
+            raise ValueError('Plese specify vertical coordinate')
+        ndim_vertical=field_in.coord_dims(vertical_axis)
+        if len(ndim_vertical)>1:
+            raise ValueError('please specify 1 dimensional vertical coordinate')
+        for index, row in features_in.iterrows():
+            if ndim_vertical[0]==0:
+                markers[:,int(row['hdim_1']), int(row['hdim_2'])]=row['feature']
+            elif ndim_vertical[0]==1:
+                markers[int(row['hdim_1']),:, int(row['hdim_2'])]=row['feature']
+            elif ndim_vertical[0]==2:
+                markers[int(row['hdim_1']), int(row['hdim_2']),:]=row['feature']
+    else:
+        raise ValueError('Segmentations routine only possible with 2 or 3 spatial dimensions')
+
+    # set markers in cells not fulfilling threshold condition to zero:
+    markers[~unmasked]=0
```

Comment:
This is the line where the example notebook `Example_Updraft_Tracking/Example_Updraft_Tracking.ipynb` breaks with an `RuntimeError: NetCDF: HDF error` issued by Xarray & Dask.

Reply:
Hmm..All four tutorials are running though without problems for me!

The versions of the probably relevant packages (should all be the newest on conda-forge) in my test environment are as follows:
xarray                    0.15.1                     py_0    conda-forge
dask                      2.14.0                     py_0    conda-forge
dask-core                 2.14.0                     py_0    conda-forge
iris                      2.4.0                    py37_0    conda-forge
netcdf4                   1.5.3           nompi_py37hdc49583_104    conda-forge
libnetcdf                 4.7.4           nompi_h84807e1_102    conda-forge
hdf5                      1.10.6          nompi_h3c11f04_100    conda-forge

Reply:
Max, I virtually have the same package version ( I created a fresh conda env for tesing).

I made a rerun of the "problematic" notebook and the error did not appear again. Sorry for this confusion!

(I did run all notebooks on my local machine on the same time - might be that I run into memory issues without noticing...) 

Hence, everything is fine - I will search for the "review accepted" button"

Cheers, Fabian

Reply:
Okay, so it was probably just some kind of hick-up no directly related to tobac... Usually when I got these kinds of errors, it was memory/storage related or related to files being acessed in parallel by two things where they shouldn't (although here each notebooks has it's separate input files...). Let's assume it's a one-off, but keep an eye open in case anyone is running into that again at some point..

Reply:
Cool, let's wait for Will's review at some point early next week (I know he's busy with other stuff this week) and then merge the pull request in as a basis for the next development steps!